### PR TITLE
fix(take): notifies errors after limit is reached

### PR DESCRIPTION
**Pull request by backbone87**
**Description:**
provides a failing test case for an edge case of take and dependent operators as presented in https://github.com/ReactiveX/rxjs/issues/5487

Will add a fix, if a change in behavior is agreed upon.

**Related issue (if exists):** fixes https://github.com/ReactiveX/rxjs/issues/5487

### CODE REVIEW

**Comment by cartant** [comment]
The behaviour reported in that issue is a bug. It's supposed to stop listening.

IMO, I would deem this a bug fix - pretty much any bug fix could be deemed a breaking change - but could be persuaded that it's a break and should go into 8.x. Either way, this has to be fixed. Broadly similar to the fix in https://github.com/ReactiveX/rxjs/pull/6151.

I have this on my of things to do/fix, but you - or anyone else - are welcome to do it. I also have noted that `elementAt`, `find` and `takeWhile` have - or might have? - a similar issue.

**Comment by benlesh** [comment]
Hmm... I just closed that issue as "not a bug". It seems more like a misunderstanding in the order of execution. I'll reopen because I'm interested in other opinions.

**Comment by backbone87** [comment]
while checking out take, i noticed that it also doesn't handle non-ints very well either.

imho, it should either check for `Number.MAX_SAFE_INTEGER` and floor the count or try to deal with all numbers (the latter can be problematic because of incrementing doesn't work past `Number.MAX_SAFE_INTEGER`

**Comment by benlesh** [comment]
@backbone87 ... probably a separate issue to address.

I do have a fix for take, though. Let's add a `skip` to this test and get it merged, and I'll have a look.

File: spec/operators/take-spec.ts [hunk 1]
```diff
@@ -217,4 +217,21 @@ describe('take', () => {
       expectSubscriptions(e1.subscriptions).toBe(e1subs);
     });
   });
+
+  it('should unsubscribe from the source when it reaches the limit before a recursive synchronous upstream error is caught', () => {
```
Comment (benlesh) [inline spec/operators/take-spec.ts:221]:
```suggestion
  it.skip('should unsubscribe from the source when it reaches the limit before a recursive synchronous upstream error is caught', () => {
```

**Comment by backbone87** [comment]
ah we pushed at the same time

**Comment by backbone87** [comment]
feel free to edit around the PR & branch, I wont push anymore now to avoid any more conflicts since @benlesh is working on it

**Comment by benlesh** [comment]
@cartant yeah, we're already handling reentrancy in the `next` case, just not in the `error` and `complete` cases. This fix is simple enough.

**Comment by benlesh** [comment]
Thanks for this, @backbone87, very helpful!

**Comment by benlesh** [comment]
Follow up PR is here: https://github.com/ReactiveX/rxjs/pull/6396
